Beam/resources/assets/js/components/IotaService.vue

```diff
@@ -1,199 +1,307 @@
-                        {minutes: 15, label: "15m", order: 0},
-                        {minutes: undefined, label: "Total", order: 1},
-                    ];
-                },
+import Axios from "axios";
+import EventHub from "./EventHub";
+
+export default {
+  name: "iota-service",
+  props: {
+    articleIds: {
+      type: Array,
+      required: true
+    },
+    articleDetailId: {
+      type: String,
+      required: false
+    },
+    pageviewMinuteRanges: {
+      type: Array,
+      default: function() {
+        return [
+          { minutes: 15, label: "15m", order: 0 },
+          { minutes: undefined, label: "Total", order: 1 }
+        ];
+      }
+    },
+    variantsMinuteRanges: {
+      type: Array,
+      default: function() {
+        return [
+          { minutes: 15, label: "15m", order: 0 },
+          { minutes: undefined, label: "Total", order: 1 }
+        ];
+      }
+    },
+    baseUrl: {
+      type: String,
+      required: true
+    },
+    configUrl: {
+      type: String,
+      required: false
+    },
+    httpHeaders: {
+      type: Object,
+      default: function() {
+        return {};
+      }
+    },
+    onArticleDetail: {
+      type: Boolean,
+      required: true
+    }
+  },
+  data: function() {
+    return {
+      config: null
+    };
+  },
+  created: function() {
+    EventHub.$on("metrics-settings-changed", this.refetchAllData);
+
+    let vm = this;
+    Object.keys(this.httpHeaders).forEach(function(name) {
+      Axios.defaults.headers.common[name] = vm.httpHeaders[name];
+    });
+    Axios.defaults.headers.common["Content-Type"] = "application/json";
+
+    new Promise((resolve, reject) => {
+      if (this.configUrl) {
+        Axios.get(this.configUrl).then(({ data }) => resolve(data), reject);
+      } else {
+        resolve();
+      }
+    }).then(config => {
+      if (config) {
+        EventHub.$emit("config-changed", config);
+      }
+
+      this.refetchAllData();
+    });
+  },
+  methods: {
+    refetchAllData(deviceType, articleLocked, subscriber) {
+      const now = new Date();
+
+      if (this.onArticleDetail) {
+        this.fetchReadProgressStats(deviceType, articleLocked);
+      }
+
+      this.fetchCommerceStats(deviceType, subscriber);
+      this.fetchPageviewStats(now, deviceType, subscriber);
+      this.fetchVariantStats(
+        now,
+        ["title_variant", "image_variant"],
+        deviceType,
+        subscriber
+      );
+    },
+    fetchCommerceStats: function(deviceType = "all", subscriber = "all") {
+      const payload = {
+        filter_by: [
+          {
+            tag: "article_id",
+            values: this.articleIds
+          },
+          {
+            tag: "derived_ua_device",
+            values:
+              deviceType === "all"
+                ? ["Computer", "Phone", "Tablet"]
+                : [deviceType]
+          }
+          // FIXME: this section, if uncommented, causes to return 0 in every conversion bubble
```

Comment:
There's a discrepancy between what we thought we tracked and what we actually track. The `subscriber` flag is only tracked for signed in users and it's missing for anonymous users. Requesting `all` as records with `true` and `false` doesn't include anonymous users. To fix this, let's alter the implementation this way:

* When `all` is requested, omit the filter for `subscriber` flag completely
* When `subscriber` is requested, use the filter as follows:

    ```json5
    {
        "filter_by": [
            // ...
            {
                "tag": "subscriber",
                "values": ["true"],
                "inverse": false
            },
        ]
    }
    ```

* When `visitor` is requested, use the filter as follows (`inverse` property has changed):

    ```json5
    {
        "filter_by": [
            // ...
            {
                "tag": "subscriber",
                "values": ["true"],
                "inverse": true
            },
        ]
    }
    ```

Similarly, let's change `deviceType` a bit too. For `all` devices omit the filter completely and use the filter only when specific device type is requested.

Btw. I actually wasn't able to reproduce zero-conversion bubbles right now, but they might be related to the filter basically omitting half of the user-base so we'll see after the changes.

Reply:
Okay, I've switched it to the new way of fetching filtered data.

One small concern in my testing:
For example there is 6357 readers with reader type filter set to 'all' (basically omitted subscriber filter).
For subscribers only I get 3553 and for visitors 2915.
Those numbers don't quite add up.
6357-3553-2915 = -111

Same thing for device type in another article:
All : 1515
Computer: 580
Tablet: 63
Phone: 870
1515-580-63-870 = 2

Reply:
I see the concern, we'll see whether it will be consistent and we'll create a ticket for checking whether we're not losing some data with this setup. In case we'll need a change, we'll create an issue in your repo.

There's also a change we'll make a feature request soon for non-article pages. In case there's no articleId, we'd need to do the page progress visualization based on the URL. We'll create an issue for that soon.

I'm going to merge this now and let some people test this for a while to get feedback and gather change requests if there are some.